src/engine/GroupByImpl.cpp

```diff
@@ -1642,3 +1675,70 @@ std::unique_ptr<Operation> GroupByImpl::cloneImpl() const {
   return std::make_unique<GroupByImpl>(_executionContext, _groupByVariables,
                                        _aliases, _subtree->clone());
 }
+
+// _____________________________________________________________________________
+// Sampling guard: skip hash-map grouping if estimated distinct groups exceed
+// threshold
+bool GroupByImpl::shouldSkipHashMapGrouping(const IdTable& table) const {
+  // Fetch runtime parameters
+  // sample size = k * sqrt(n)
+  if (!RuntimeParameters().get<"group-by-sample-enabled">()) {
+    return false;
+  }
+  size_t maxSampleRows = RuntimeParameters().get<"group-by-sample-max-rows">();
+  size_t groupThreshold =
+      RuntimeParameters().get<"group-by-sample-group-threshold">();
+  double distinctRatio =
+      RuntimeParameters().get<"group-by-sample-distinct-ratio">();
+  size_t totalSize = table.size();
+  size_t k = RuntimeParameters().get<"group-by-sample-constant">();
+
+  size_t sampleSize = k * static_cast<size_t>(std::sqrt(double(totalSize)));
+  if (sampleSize == 0) {
+    return false;
+  }
+  if (maxSampleRows != 0) {
+    sampleSize = std::min(sampleSize, maxSampleRows);
+  }
+  // Reservoir-sample indices only
+  std::vector<size_t> indices(sampleSize);
+  for (size_t i = 0; i < sampleSize; ++i) indices[i] = i;
+  std::mt19937_64 gen{42};
+  for (size_t i = sampleSize; i < totalSize; ++i) {
+    std::uniform_int_distribution<size_t> dist(0, i);
+    size_t randIdx = dist(gen);
+    if (randIdx < sampleSize) {
+      indices[randIdx] = i;
+    }
+  }
+  // Use absl::flat_hash_map to count frequencies for Chao1 estimator
+  absl::flat_hash_map<RowKey, size_t> groupCounts;
+  const auto& varCols = _subtree->getVariableColumns();
+  std::vector<ColumnIndex> groupByCols;
+  groupByCols.reserve(_groupByVariables.size());
+  for (const auto& var : _groupByVariables) {
+    groupByCols.push_back(varCols.at(var).columnIndex_);
+  }
+  for (size_t idx : indices) {
+    RowKey key{&table, idx, &groupByCols};
+    ++groupCounts[key];
+  }
```

Comment:
You can do this directly for each of the sampled indices, no need to materialized them in a vector first.